### PR TITLE
Add CSG unit and unit builder

### PR DESCRIPTION
**Pull request by sethrj**
This is the next layer in scene construction: a helper class for the upcoming object builder that manages the components of a new unit, tying together surfaces, metadata, transforms, CSG nodes, volumes, and volume fills. The `CsgUnit` will be built up by shapes, solids, etc.; and a lot of the code in this PR is for efficiently testing that constructed object.

### CODE REVIEW

**Comment by sethrj** [comment]
Ping @elliottbiondo

File: src/orange/orangeinp/detail/CsgUnitBuilder.hh [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    //!@{
+    //! \name Type aliases
+    using Tol = Tolerance<>;
+    using Metadata = CsgUnit::Metadata;
+    using NodeId = CsgUnit::NodeId;
+    using BBox = CsgUnit::BBox;
+    //!@}
+
+  public:
+    // Construct with a "processed unit" to build.
```
Comment (elliottbiondo) [inline src/orange/orangeinp/detail/CsgUnitBuilder.hh]: This be a "fresh unit", per the function definition below

File: src/orange/orangeinp/detail/CsgUnitBuilder.hh [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with a fresh unit.
+ */
+CsgUnitBuilder::CsgUnitBuilder(CsgUnit* u, Tolerance<> const& tol)
+    : unit_{u}, tol_{tol}, insert_surface_{&unit_->surfaces, tol}
+{
+    CELER_EXPECT(unit_);
+    CELER_EXPECT(unit_->metadata.size() == 0 && unit_->bboxes.size() == 0
```
Comment (elliottbiondo) [inline src/orange/orangeinp/detail/CsgUnitBuilder.hh]: Maybe `CsgUnit` could have an `is_fresh` or `is_empty` method to replace this line

File: src/orange/orangeinp/detail/CsgUnitBuilder.cc [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+namespace celeritas
+{
+namespace orangeinp
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Mark a CSG node as a volume of real space.
```
Comment (elliottbiondo) [inline src/orange/orangeinp/detail/CsgUnitBuilder.cc]: Fix doc string

**Review by elliottbiondo — APPROVED** [review]

**Pull request merged**